miner/proposal_builder.go

```diff
@@ -374,137 +335,370 @@ func (pb *ProposalBuilder) decideMeshHash(ctx context.Context, current types.Lay
+			Ballot: types.Ballot{
+				InnerBallot: types.InnerBallot{
+					Layer:       lid,
+					AtxID:       session.atx,
+					OpinionHash: opinion.Hash,
+				},
+				Votes:             opinion.Votes,
+				EligibilityProofs: eligibility,
+			},
+			TxIDs:    txs,
+			MeshHash: meshHash,
+		},
+	}
+	if session.ref == types.EmptyBallotID {
+		p.Ballot.RefBallot = types.EmptyBallotID
+		p.Ballot.EpochData = &types.EpochData{
+			ActiveSetHash:    session.active.set.Hash(),
+			Beacon:           session.beacon,
+			EligibilityCount: session.eligibilities.slots,
+		}
+	} else {
+		p.Ballot.RefBallot = session.ref
+	}
+	p.Ballot.Signature = signer.Sign(signing.BALLOT, p.Ballot.SignedBytes())
+	p.SmesherID = signer.NodeID()
+	p.Signature = signer.Sign(signing.PROPOSAL, p.SignedBytes())
+	p.MustInitialize()
+	return p
+}
 
-	txList := pb.conState.SelectProposalTXs(layerID, len(proofs))
-	p, err := pb.createProposal(ctx, layerID, epochEligibility, beacon, txList, *opinion)
+func ActiveSetFromEpochFirstBlock(db sql.Executor, epoch types.EpochID) ([]types.ATXID, error) {
+	bid, err := layers.FirstAppliedInEpoch(db, epoch)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("first block in epoch %d not found: %w", epoch, err)
 	}
+	return activeSetFromBlock(db, bid)
+}
 
-	pb.saveMetrics(ctx, started, layerID)
+func activeSetFromBlock(db sql.Executor, bid types.BlockID) ([]types.ATXID, error) {
+	block, err := blocks.Get(db, bid)
+	if err != nil {
+		return nil, fmt.Errorf("actives get block: %w", err)
+	}
+	activeMap := make(map[types.ATXID]struct{})
+	// the active set is the union of all active sets recorded in rewarded miners' ref ballot
+	for _, r := range block.Rewards {
+		activeMap[r.AtxID] = struct{}{}
+		ballot, err := ballots.FirstInEpoch(db, r.AtxID, block.LayerIndex.GetEpoch())
+		if err != nil {
+			return nil, fmt.Errorf("actives get ballot: %w", err)
+		}
+		actives, err := activesets.Get(db, ballot.EpochData.ActiveSetHash)
+		if err != nil {
+			return nil, fmt.Errorf("actives get active hash for ballot %s: %w", ballot.ID().String(), err)
+		}
+		for _, id := range actives.Set {
+			activeMap[id] = struct{}{}
+		}
+	}
+	return maps.Keys(activeMap), nil
+}
 
-	if pb.stopped() {
-		return nil
+func activesFromFirstBlock(
+	cdb *datastore.CachedDB,
+	signer *signing.VRFSigner,
+	target types.EpochID,
+	ownAtx types.ATXID,
+	ownWeight uint64,
+) (uint64, []types.ATXID, error) {
+	set, err := ActiveSetFromEpochFirstBlock(cdb, target)
+	if err != nil {
+		return 0, nil, err
+	}
+	var (
+		totalWeight uint64
+		ownIncluded bool
+	)
+	for _, id := range set {
+		ownIncluded = ownIncluded || id == ownAtx
+		atx, err := cdb.GetAtxHeader(id)
+		if err != nil {
+			return 0, nil, err
+		}
+		totalWeight += atx.GetWeight()
+	}
+	if !ownIncluded {
+		// miner is not included in the active set derived from the epoch's first block
+		set = append(set, ownAtx)
+		totalWeight += ownWeight
 	}
+	return totalWeight, set, nil
```

Comment:
It could be simpler if `ActiveSetFromEpochFirstBlock` returned a set (a `map[types.ATXID]struct{}`) - you could just insert `ownAtx` w/o checking if its already there.

Reply:
i am changing this in a followup, as i will not forcefully include my own atx. wanted to keep behavior the same in this pr